lib/Conversion/TritonGPUToLLVM/Utility.cpp

```diff
@@ -316,6 +308,21 @@ static Value commonShflSync(Location loc, ConversionPatternRewriter &rewriter,
 
 #ifdef USE_ROCM
   GCNBuilder builder;
+
+  auto permute = [&](Value lineId, StringRef permuteInstStr) {
+    assert(permuteInstStr == "ds_permute_b32" ||
+           permuteInstStr == "ds_bpermute_b32");
+    // Multiple lineId by 4. (More on permute instruction semantics:
+    // https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/instruction-set-architectures/instinct-mi200-cdna2-instruction-set-architecture.pdf#page=180
+    Value byteOffset = i32_val(2);
+    Value permuteAddr = shl(lineId, byteOffset);
+    auto shfl = builder.create(permuteInstStr.str());
+    auto dOpr = builder.newOperand("=v");
+    auto addrOpr = builder.newOperand(permuteAddr, "v");
+    auto aOpr = builder.newOperand(val, "v");
+    (*shfl)(dOpr, addrOpr, aOpr);
```

Comment:
Note: `rocdl` dialect recently got `bpermute` operation.
We can replace this inline asm someday.

Reply:
oh yea? I was looking for various shuffles in rocdl, but I didn't see them - I guess I should have been looking for permute instead. That's good.

Reply:
@binarman Do you happen to have the commit in MLIR?

Reply:
https://github.com/llvm/llvm-project/commit/1896096002b75b50d46ee0043c20e90c7e27604a